lib/ddtrace/contrib/ethon/easy_patch.rb

```diff
@@ -0,0 +1,146 @@
+      # Ethon EasyPatch
+      module EasyPatch
+        def self.included(base)
+          if Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.0.0')
+            base.class_eval do
+              # Among the patched methods, only this one is defined on Easy class, others are included
+              alias_method :set_attributes_without_datadog, :set_attributes
+              remove_method :set_attributes
+
+              include InstanceMethods
+            end
+          else
+            base.send(:prepend, InstanceMethods)
+          end
+        end
+
+        # Compatibility shim for Rubies not supporting `.prepend`
+        module InstanceMethodsCompatibility
+          def set_attributes(options)
+            set_attributes_without_datadog(options)
+          end
+        end
+
+        # InstanceMethods - implementing instrumentation
+        module InstanceMethods
+          include InstanceMethodsCompatibility unless Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')
+
+          def http_request(url, action_name, options = {})
+            return super unless tracer_enabled?
+
+            # It's tricky to get HTTP method from libcurl
+            @datadog_method = action_name.to_s.upcase
+            super
+          end
+
+          def set_attributes(options)
+            return super unless tracer_enabled?
+
+            # Make sure headers= will get called
+            options[:headers] ||= {}
+            super options
+          end
+
+          def headers=(headers)
+            return super unless tracer_enabled?
+
+            # Store headers to call this method again when span is ready
+            headers ||= {}
+            @datadog_original_headers = headers
+            super headers
+          end
+
+          def perform
+            return super unless tracer_enabled?
+            datadog_before_request
+            super
+          end
+
+          def complete
+            return super unless tracer_enabled?
+            begin
+              response_options = mirror.options
+              response_code = (response_options[:response_code] || response_options[:code]).to_i
+              return_code = response_options[:return_code]
+              if response_code.zero?
+                message = return_code ? ::Ethon::Curl.easy_strerror(return_code) : 'unknown reason'
+                set_span_error_message("Request has failed: #{message}")
+              else
+                @datadog_span.set_tag(Datadog::Ext::HTTP::STATUS_CODE, response_code)
+                if Datadog::Ext::HTTP::ERROR_RANGE.cover?(response_code)
+                  set_span_error_message("Request has failed with HTTP error: #{response_code}")
+                end
+              end
+            ensure
+              @datadog_span.finish
+              @datadog_span = nil
+            end
+            super
+          end
+
+          def datadog_before_request
+            @datadog_span = datadog_configuration[:tracer].trace(
```

Comment:
What's the thought for using an instance variable here? It looks like maybe we need something to work for sync and async requests?

How does this work in a multithreaded application, or one that makes multiple async calls? My thought when I saw the use of an instance variable for this was: "can this conflict with other Ethon requests that are being made in parallel? This state would presumably be shared across all requests through the same instance of Ethon."

Reply:
Okay, I think that's the key part; we don't leave unfinished spans (which will break traces), and we don't overwrite or accidentally re-use these spans for different operations.